content/en/blog/2023/2023-03-08-scheduler-plugin-reusing/index.md

```diff
@@ -0,0 +1,161 @@
+		    f.frameworkHandle, f.topologyKey)
+		// calculate score between 0-100 linearly based on distance from min and max
+		scores[i].Score := linearize(currentNodeTopologyScoreSum, minTopologyScoreSum, maxTopologyScoreSum)
+	}
+
+	return nil
+}
+```
+
+With `NodeResourcesFit` as the piped plugin, such a `TopologyScorer` could give scores based on overall resource status of the rack the nodes belong to, assuming a rack topology key would be used in configuring the `TopologyScorer`.
+
+Even such a simple graph can reveal a lot of potential over simple composition. Namely:
+ * It doesn't run node resource scoring twice
+ * It doesn't require configuring node resource scoring for two plugins
+ * It only depends on the reused plugin being able to provide node scores. Internal changes to the reused plugin won't break the plugin which reuses the results, as long as scores are created.
+ * It allows changing the connected node score source plugin with a single value in plugin args. What used to be a hard-coded topology resource scorer composite, became a generic TopologyScorer which can use any other score plugin as the source. There are no changes needed to the reused plugins.
+
+This sort of plugin connecting would allow creating simple, quickly running plugins which modify the node scores of the connected source plugin. For instance, a plugin which optionally normalizes and/or inverts (reverses) the scores of the connected source plugin, would be rather
+trivial to create. The new plugin could choose to either alter the scores of the source, or output its own scores. The author experimented with multiple daisy-chained plugins which altered the original source plugin scores, and it worked fine. The obvious caveat lies in the plugins being single instances. It would be easy to normalize, reverse or otherwise change the scores of one plugin, but what if one would need to do it for two plugins? Solving that would require further innovating.
+
+## Summary
+
+There are several ways to do Kubernetes scheduler plugin reusing, ranging from strictly not recommended copy-pasting to composition and all the way to modifying the scheduler framework, depending on how efficient you need things to be. The scheduler framework includes built-in utilities such as the `CycleState`, and by no means does it prevent plugins from passing information between each other. The pieces are there, only thing currently missing is perhaps a bit of assisting from the side of the Scheduler framework runtime.
+
+## Next steps
+
+As per discussed in the sig-scheduling meetings, the topology resource scoring plugin will proceed with the more conservative composition approach, because it can get the job done without framework changes, associated risks and maintenance burden. However, if you find that it would indeed benefit your use-cases to be able to simply configure scheduler profiles in such a way that certain score plugins are connected, chime in at slack [#sig-scheduling].
```

Comment:
```suggestion
As per discussed in SIG Scheduling meetings, the topology resource scoring plugin will proceed with the more conservative composition approach, because it can get the job done without framework changes, associated risks and maintenance burden. However, if you find that it would indeed benefit your use-cases to be able to simply configure scheduler profiles in such a way that certain score plugins are connected, chime in via [`#sig-scheduling`][#sig-scheduling] on Kubernetes Slack (visit https://slack.k8s.io/ for an invitation).
```